src/Compiler/Opts/CSE.idr

```diff
@@ -199,159 +198,291 @@ mutual
   analyze c@(CForce _ _ _)   = analyzeSubExp c
   analyze c@(CDelay _ _ _)   = analyzeSubExp c
 
-  analyze exp = case dropEnv {pre = []} exp of
-    Just e0 => do
-      count <- store e0
-      -- only analyze subexpressions of closed terms once
-      when (count == 1) $ analyzeSubExp exp
-    Nothing => analyzeSubExp exp
-
-  analyzeSubExp : Ref Sts St => CExp ns -> Core ()
-  analyzeSubExp (CLocal _ _)         = pure ()
-  analyzeSubExp (CRef _ _)           = pure ()
-  analyzeSubExp (CLam _ _ y)         = analyze y
-  analyzeSubExp (CLet _ _ _ y z)     = analyze y >> analyze z
-  analyzeSubExp (CApp fc x xs)       = analyze x >> traverse_ analyze xs
-  analyzeSubExp (CCon _ _ _ _ [])    = pure ()
-  analyzeSubExp (CCon _ _ _ _ xs)    = traverse_ analyze xs
-  analyzeSubExp (COp _ _ xs)         = ignore $ traverseVect analyze xs
-  analyzeSubExp (CExtPrim _ _ xs)    = traverse_ analyze xs
-  analyzeSubExp (CForce _ _ y)       = analyze y
-  analyzeSubExp (CDelay _ _ y)       = analyze y
-  analyzeSubExp (CConCase _ sc xs x) =
-    analyze sc                     >>
-    traverse_ analyzeConAlt xs     >>
-    ignore (traverseOpt analyze x)
-
-  analyzeSubExp (CConstCase _ sc xs x) =
-    analyze sc                     >>
-    traverse_ analyzeConstAlt xs   >>
-    ignore (traverseOpt analyze x)
-
-  analyzeSubExp (CPrimVal _ _) = pure ()
-  analyzeSubExp (CErased _)    = pure ()
-  analyzeSubExp (CCrash _ _)   = pure ()
-
-  analyzeConAlt : { auto c : Ref Sts St } -> CConAlt ns -> Core ()
-  analyzeConAlt (MkConAlt _ _ _ _ z) = analyze z
-
-  analyzeConstAlt : Ref Sts St => CConstAlt ns -> Core ()
-  analyzeConstAlt (MkConstAlt _ y) = analyze y
-
-analyzeDef : Ref Sts St => CDef -> Core ()
-analyzeDef (MkFun args x)    = analyze x
-analyzeDef (MkCon _ _ _)     = pure ()
-analyzeDef (MkForeign _ _ _) = pure ()
-analyzeDef (MkError _)       = pure ()
-
-analyzeName : Ref Sts St => Ref Ctxt Defs => Name -> Core ()
+  analyze exp = do
+    (sze, exp') <- analyzeSubExp exp
+    case dropEnv {pre = []} exp' of
```

Comment:
Isn't this still exponential? You've already analysed the sub expressions which
means you've already attempted to run `dopEnv` on them, doesn't it?

Reply:
Yes, that's why I wrote "linear space" not "linear time", but I think it is polymorphic (O(n^2)) not exponential? However, I will give this some more thought, since we might be able to make use of the info from subterm analysis and be able to get rid of the call to `dropEnv` on parent terms.

The problem with quadratic space was that it bites us even in the face of large nested closed terms. The quadratic time above only bites us in the face of large nested open terms, since closed child terms will be replace by machine generated names before the call to `dropEnv`.

Reply:
I think it could be linear in both if `analyze` returned the subterm with free variables removed as it goes.
This has the disadvantage of having a more complex function, but I think it would be worth it, seeing how much CSE helps with executable size.
eg
```idris
analyse : Ref Sts St => {locals : _} -> CExp (locals :: vars) -> Core (Maybe (CExp locals))
analyse (CLocal fc p) = pure $ CLocal fc <$> dropEnv p
analyse (CApp fc fn as) = do
    fn' <- analyse fn
    as' <- traverse analyse as
    pure $ CApp <$> fn' <*> as'
...
```

Reply:
> I think it could be linear in both if `analyze` returned the subterm with free variables removed as it goes.
> This has the disadvantage of having a more complex function, but I think it would be worth it, seeing how much CSE helps with executable size.

I experimented with this, an I think it is not as simple as it seems from your example. The hard cases are those, which introduce new variables: Let bindings and lambdas. For instance, what do we do in case of `CLam _ n y`? There are three possibilities to consider:

1. `y` is a closed term (`n` makes no appearance in it), in which case it should be subjected to CSE as should the lambda itself, since it will also be a closed term.
2. `y` has only `n` as a free variable, in which case the lambda is still a closed term and should be subjected to CSE but not so `y`.
3. `y` has additional free variables, in which case neither  `y` nor the lambda should be considered during CSE.

So, in order to stick to the linear algorithm, we need to consider three possibilities in the return type of `analyze`, and these will start to multiply in the face of recursion. If - however - we invoke `analyze` twice with different lists of name prefixes,
we get exponential complexity in the case of nested lambdas (or let bindings).

The same goes for `let` bindings, which also introduce a new variable. Here we could argue that it's unlikely that the new variable is not used in the subexpression, but lambdas ignoring parts of their arguments are pretty common.

A linear version is not lost, since we could always just return the set of free variables (collected in order of appearance) together with every subterm and inspect those sets at runtime to separate closed terms from open ones (actually, this would be O (n * log n) if we use a `SortedSet` to store the names of free variables). However, I could not yet come up with a nice encoding for this, which also behaves well at compile time. Will have to think some more.